### PR TITLE
CRM-16734 participant tokens in participant pdf

### PR DESCRIPTION
**Pull request by johanv**
Continued @mallezie's work for CRM-16734 (participant tokens in PDF letters for participants). This pull request is a continuation of #8258, which provides event tokens for PDF letters to participants (CRM-16599). Now you can create PDF letters like this:

> Hi {contact.first_name},
> 
> This is your letter. You subscribed for {event.title}; the event fee amount was {participant.participant_fee_amount}.

---
- [CRM-16734: Add participant tokens to participant pdf action](https://issues.civicrm.org/jira/browse/CRM-16734)
- [CRM-16599: Add event tokens to participant pdf creation](https://issues.civicrm.org/jira/browse/CRM-16599)

### CODE REVIEW

File: CRM/Utils/Token.php [hunk 1]
```diff
@@ -1759,6 +1796,147 @@ public static function replaceMultipleContributionTokens($separator, $str, &$con
...
+      throw new Exception('$entity is expected to be "event".');
+    }
+
+    self::_buildEventTokens();
+
+    $params = array('entity_id' => $event['id'], 'entity_table' => 'civicrm_event');
+
+    switch ($token) {
+      case 'balance':
+        $info = CRM_Contribute_BAO_Contribution::getPaymentInfo($params['participant_id'], 'event');
```
Comment (agh1) [inline CRM/Utils/Token.php]: There is no key `participant_id` in the `$params` array.

File: CRM/Utils/Token.php [hunk 2]
```diff
@@ -1759,6 +1796,147 @@ public static function replaceMultipleContributionTokens($separator, $str, &$con
...
+        $value = $balancePay;
+        break;
+
+      case 'title':
+        $value = $event['event_title'];
+        break;
+
+      case 'end_date':
+      case 'start_date':
+        $value = CRM_Utils_Date::customFormat($event[$token], "%d/%m/%Y");
```
Comment (agh1) [inline CRM/Utils/Token.php]: Date format should not be hard-coded

File: CRM/Utils/Token.php [hunk 3]
```diff
@@ -1759,6 +1796,147 @@ public static function replaceMultipleContributionTokens($separator, $str, &$con
...
+          }
+        }
+        else {
+          $value = $participant['participant_fee_level'];
+        }
+        break;
+
+      case 'event_end_date':
+      case 'event_start_date':
+        $value = CRM_Utils_Date::customFormat($participant[$token], "%d/%m/%Y");
```
Comment (agh1) [inline CRM/Utils/Token.php]: Date format should not be hard-coded

File: CRM/Utils/Token.php [hunk 4]
```diff
@@ -1606,6 +1605,44 @@ protected static function _buildMembershipTokens() {
...
+   */
+  protected static function _buildEventTokens() {
+    $key = 'event';
+    if (!isset(self::$_tokens[$key]) || self::$_tokens[$key] == NULL) {
+      $eventTokens = array();
+      $tokens = CRM_Core_SelectValues::eventTokens();
+      foreach ($tokens as $token => $dontCare) {
+        $eventTokens[] = substr($token, (strpos($token, '.') + 1), -1);
+      }
+      $customtokens = CRM_CORE_BAO_CustomField::getFields('Event');
```
Comment (agh1) [inline CRM/Utils/Token.php]: Pretty sure it needs to be `CRM_Core_BAO_CustomField` (i.e. `Core` in mixed case).

File: CRM/Utils/Token.php [hunk 5]
```diff
@@ -1606,6 +1605,44 @@ protected static function _buildMembershipTokens() {
...
+   */
+  protected static function _buildParticipantTokens() {
+    $key = 'participant';
+    if (!isset(self::$_tokens[$key]) || self::$_tokens[$key] == NULL) {
+      $participantTokens = array();
+      $tokens = CRM_Core_SelectValues::participantTokens();
+      foreach ($tokens as $token => $dontCare) {
+        $participantTokens[] = substr($token, (strpos($token, '.') + 1), -1);
+      }
+      $customtokens = CRM_CORE_BAO_CustomField::getFields('Participant');
```
Comment (agh1) [inline CRM/Utils/Token.php]: Pretty sure it needs to be `CRM_Core_BAO_CustomField` (i.e. `Core` in mixed case).

File: CRM/Event/Form/Task/PDF.php [hunk 6]
```diff
@@ -99,6 +96,21 @@ public function setDefaultValues() {
    */
   public function listTokens() {
     $tokens = CRM_Core_SelectValues::contactTokens();
+    $tokens = array_merge(CRM_Core_SelectValues::eventTokens(), $tokens);
+    // unset contact_email and contact_phone tokens.
+    // These are location_email and location_contact
+    // Should be cleaned up after ActionSchedule token replacement cleanup.
+    unset($tokens['{event.contact_email}']);
+    unset($tokens['{event.contact_phone}']);
+    $customEventTokens = CRM_CORE_BAO_CustomField::getFields('Event');
```
Comment (agh1) [inline CRM/Event/Form/Task/PDF.php]: Pretty sure it needs to be `CRM_Core_BAO_CustomField` (i.e. `Core` in mixed case).

**Comment by agh1** [comment]
Thanks @johanv and @mallezie for all of this.  This looks super-promising.

Unfortunately, it has some buggy things that should be worked out.  I'm happy to help with these, but I don't think this will be ready to go for 4.7.8.  Some things to address:

- The `{participant.participant_status_id}` token gives you an integer, and there's no way to get the actual role label.
- The `{participant.participant_register_date}` token is in the format `YYYY-MM-DD HH:MM:SS`.  While different people might want different formats, that is probably the one format that nobody wants.
- `{participant.participant_fee_amount}` could be combined with `{participant.currency}`, but CiviCRM has better ways to localize the amount.  Currently, the token just gives you a number.
- ` {event.location}` is presented all in one line.  I wonder if this could be broken into multiple tokens.
- `{event.fee_amount}` is blank, at least in all of my examples.  I notice [a comment](https://issues.civicrm.org/jira/browse/CRM-16599?focusedCommentId=88264&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-88264) that it was removed: I wonder if it just needs to be excluded from the tokens list.
- `{event.balance}` causes a fatal error (due to the issue in the note above).  Anyway, the balance will be per-participant.
- The miscellaneous small things in the inline notes should be resolved.

Looking through this, I feel like this and other token use cases need test coverage.  It doesn't appear that there's anything at the moment.

**Comment by agh1** [comment]
One separate note: I don't think this has anything to do with this issue, but I was getting an error trying to generate a PDF at all because vendor/dompdf/dompdf/lib/html5lib/Parser.php requires vendor/dompdf/dompdf/lib/html5lib/TreeBuilder.php, but TreeBuilder.php is deleted in [tools/scripts/composer/dompdf-cleanup.sh](https://github.com/civicrm/civicrm-core/blob/master/tools/scripts/composer/dompdf-cleanup.sh#L117).  Commenting out the line in Parser.php that requires TreeBuilder.php makes everything work just fine, however.

**Comment by totten** [comment]
@agh1 @xurizaemon - the `TreeBuilder.php` comment seems like a recent regression from #7991. Since it's tangential to this PR, suggest any further follow-up on `TreeBuilder.php` be on proposed fix #8419.

**Comment by eileenmcnaughton** [comment]
@agh1 Can you check this again?

**Comment by jitendrapurohit** [comment]
While working on QA for this PR, I found some minor issues related to the above changes after the update-

1) Use of `participant.role`, `participant.role_id` tokens is displayed as `Array` and not the values for it.
2) The `{participant.participant_status_id}` token gives you an integer. It should actually display as label.
3) `participant.transferred_to_contact_id` token doesn't seem to be working.
4) `event.fee_amount` is blank as @agh1 mentioned.
4) Found some set of e-notices - http://goo.gl/iD1C8P

**Comment by johanv** [comment]
I will look at @jitendrapurohit 's comments, probably next week, but i will do some cleanup in this code first.

**Comment by eileenmcnaughton** [comment]
It would be great if you would do a rebase -i on this & squash some patches

**Comment by johanv** [comment]
This PR is replaced by #9038.

**Pull request closed without merging**